toxcore/announce.c

```diff
@@ -0,0 +1,649 @@
+
+        if (entry->data == nullptr) {
+            return false;
+        }
+
+        memcpy(entry->data, data, length);
+    }
+
+    entry->length = length;
+    memcpy(entry->data_public_key, data_public_key, CRYPTO_PUBLIC_KEY_SIZE);
+    entry->store_until = mono_time_get(announce->mono_time) + timeout;
+
+    return true;
+}
+
+non_null()
+static uint32_t calculate_timeout(const Announcements *announce, uint32_t requested_timeout)
+{
+    const uint64_t uptime = mono_time_get(announce->mono_time) - announce->start_time;
+    const uint32_t max_announcement_timeout = max_u32(
+                (uint32_t)min_u64(
+                    MAX_MAX_ANNOUNCEMENT_TIMEOUT,
+                    uptime / MAX_ANNOUNCEMENT_TIMEOUT_UPTIME_RATIO),
+                MIN_MAX_ANNOUNCEMENT_TIMEOUT);
+
+    return min_u32(max_announcement_timeout, requested_timeout);
+}
+
+#define DATA_SEARCH_TO_AUTH_MAX_SIZE (CRYPTO_PUBLIC_KEY_SIZE * 2 + MAX_PACKED_IPPORT_SIZE + MAX_SENDBACK_SIZE)
+
+non_null(1, 2, 3, 4, 7) nullable(5)
+static int create_data_search_to_auth(const Logger *logger, const uint8_t *data_public_key,
+                                      const uint8_t *requester_key,
+                                      const IP_Port *source, const uint8_t *sendback, uint16_t sendback_length,
+                                      uint8_t *dest, uint16_t max_length)
+{
+    if (max_length < DATA_SEARCH_TO_AUTH_MAX_SIZE) {
+        return -1;
+    }
+
+    memcpy(dest, data_public_key, CRYPTO_PUBLIC_KEY_SIZE);
+    memcpy(dest + CRYPTO_PUBLIC_KEY_SIZE, requester_key, CRYPTO_PUBLIC_KEY_SIZE);
+
+    const int ipport_length = pack_ip_port(logger, dest + CRYPTO_PUBLIC_KEY_SIZE * 2, MAX_PACKED_IPPORT_SIZE, source);
+
+    if (ipport_length == -1) {
+        return -1;
+    }
+
+    if (sendback_length > 0) {
+        assert(sendback != nullptr);
+        memcpy(dest + CRYPTO_PUBLIC_KEY_SIZE * 2 + ipport_length, sendback, sendback_length);
+    }
+
+    return CRYPTO_PUBLIC_KEY_SIZE * 2 + ipport_length + sendback_length;
+}
+
+#define DATA_SEARCH_TIMEOUT 60
+
+non_null(1, 2, 3, 7, 9) nullable(5)
+static int create_reply_plain(Announcements *announce,
+                              const uint8_t *requester_key, const IP_Port *source, uint8_t type,
+                              const uint8_t *sendback, uint16_t sendback_length,
+                              const uint8_t *data, uint16_t length,
+                              uint8_t *reply, uint16_t reply_max_length)
+{
+    if (length < CRYPTO_PUBLIC_KEY_SIZE) {
+        return -1;
+    }
+
+    const uint8_t *const data_public_key = data;
+
+    uint8_t to_auth[DATA_SEARCH_TO_AUTH_MAX_SIZE];
+    const int to_auth_length = create_data_search_to_auth(announce->log, data_public_key, requester_key, source,
+                               sendback, sendback_length, to_auth, DATA_SEARCH_TO_AUTH_MAX_SIZE);
+
+    if (to_auth_length == -1) {
+        return -1;
+    }
+
+    if (type == NET_PACKET_DATA_SEARCH_REQUEST) {
```

Comment:
This looks like a `switch`. Also, it's probably worth extracting functions for each of the cases, as they are quite large.